src/main/java/org/wildfly/security/sasl/util/AuthenticationTimeoutSaslServerFactory.java

```diff
@@ -0,0 +1,126 @@
+     */
+    public static final long DEFAULT_TIMEOUT = 150;
+
+    private final ScheduledExecutorService scheduledExecutorService;
+
+    /**
+     * Construct a new instance.
+     *
+     * @param delegate the delegate {@code SaslServerFactory}
+     * @param scheduledExecutorService the scheduled executor to use to handle authentication timeout tasks
+     */
+    public AuthenticationTimeoutSaslServerFactory(final SaslServerFactory delegate, final ScheduledExecutorService scheduledExecutorService) {
+        super(delegate);
+        Assert.checkNotNullParam("scheduledExecutorService", scheduledExecutorService);
+        this.scheduledExecutorService = scheduledExecutorService;
+    }
+
+    @Override
+    public SaslServer createSaslServer(final String mechanism, final String protocol, final String serverName, final Map<String, ?> props, final CallbackHandler cbh) throws SaslException {
+        final long timeout;
+        if (props.containsKey(WildFlySasl.AUTHENTICATION_TIMEOUT)) {
+            timeout = Long.parseLong((String) props.get(WildFlySasl.AUTHENTICATION_TIMEOUT));
+        } else {
+            timeout = DEFAULT_TIMEOUT;
+        }
+
+        final SaslServer delegateSaslServer = delegate.createSaslServer(mechanism, protocol, serverName, props, cbh);
+        return (delegateSaslServer == null) ? null : new DelegatingTimeoutSaslServer(delegateSaslServer, scheduledExecutorService, timeout);
+    }
+
+    private static class DelegatingTimeoutSaslServer extends AbstractDelegatingSaslServer {
+        private final AtomicBoolean complete = new AtomicBoolean();
+        private volatile ScheduledFuture<Void> timeoutTask;
+
+        DelegatingTimeoutSaslServer(final SaslServer delegate, final ScheduledExecutorService scheduledExecutorService, final long timeout) {
+            super(delegate);
+
+            // Schedule a task to terminate the authentication attempt if it takes too long
+            timeoutTask = scheduledExecutorService.schedule(() -> {
+                if (! (isComplete() && complete.compareAndSet(false, true))) {
+                    dispose();
+                }
+                return null;
+            }, timeout, TimeUnit.SECONDS);
+        }
+
+        @Override
+        public byte[] evaluateResponse(final byte[] response) throws SaslException {
+            try {
+                final byte[] challenge = delegate.evaluateResponse(response);
+                if (isComplete() && complete.compareAndSet(false, true)) {
+                    cancelTimeoutTask();
+                }
+                return challenge;
+            } catch (SaslException | RuntimeException | Error e) {
+                if (isComplete() && complete.compareAndSet(false, true)) {
+                    cancelTimeoutTask();
+                }
+                throw e;
+            }
+        }
+
+        @Override
+        public void dispose() throws SaslException {
+            try {
```

Comment:
While technically this doesn't matter from a correctness perspective if there is a synchronized block protecting access to the delegate and timeout task, the timeout task _should_ be cancelled first in case dispose() is unusually expensive to avoid possible spurious task execution which would block the task thread pointlessly for some amount of time.